### DISCUSSION TITLE
GridSplitter can no longer be used to create a subclass

### DISCUSSION THREAD

**Discussion by kuiperzone** [AvaloniaUI/Avalonia#12989] [Q&A]
I have a CustomSplitter class derived from GridSplitter to provide custom behaviour. This has always worked but stopped working around 11.0.3.

It builds and runs, but the custom splitter is now totally unresponsive to mouse input.

To clarify, if I create a trivial subclass:

  public class CustomSplitter : GridSplitter
  {
    public CustomSplitter()
      : base()
    {
    }
  }

and reference it in XAML, it is not responsive at all to mouse input.

This used to work? Any reason?

Running Linux.

**Describe the bug**

**To Reproduce**

**Expected behavior**

**Screenshots**

**Desktop (please complete the following information):**

- OS: [e.g. Windows, Mac, Linux (State distribution)]
- Version [e.g. 0.10.0-rc1 or 0.9.12]

**Additional context**

**Comment by timunie** [comment]
Works for me in 11.0.4, if I apply the needed style. As I have no idea if you use a custom style, I re-used the original GridSplitter-Style: 

```diff
  public class CustomSplitter : GridSplitter
  {
      public CustomSplitter()
          : base()
      {
      }

+     protected override Type StyleKeyOverride => typeof(GridSplitter);
  }
```

Please double-check the used Style / ControlTheme.

**Comment by kuiperzone** [comment]
> `protected override Type StyleKeyOverride => typeof(GridSplitter);`

Thanks for that!

I applied the same thing to the custom TextBox control I have a problem with and that bit works now. (Ref https://github.com/AvaloniaUI/Avalonia/issues/12967)

However, it's not working for me with the CustomSplitter (not yet anyway). Will look into it more and update and close if applicable.

Thank you!

**Comment by kuiperzone** [comment]
PS. You just made me happy again. :)

**Comment by kuiperzone** [comment]
Applying the recommendations to my other control worked. Ref: https://github.com/AvaloniaUI/Avalonia/issues/12967

However, there's still a problem with `GridSplitter`. The following simple class with `StyleKeyOverride` does not appear to work and the splitter is unresponsive. Any thoughts?

```
public class CustomSplitter : GridSplitter
{
    private readonly DispatcherTimer _timer;
    private IBrush? _originalBackground;

    /// <summary>
    /// Constructor.
    /// </summary>
    public CustomSplitter()
    {
        _timer = new(TimeSpan.FromMilliseconds(250), DispatcherPriority.Normal, TimerHandler);
        PointerMoved += PointerMovedHandler;
    }

    /// <summary>
    /// Custom property.
    /// </summary>
    public static readonly StyledProperty<IBrush?> HighlightProperty =
        AvaloniaProperty.Register<CustomSplitter, IBrush?>(nameof(Highlight), Brushes.Gray);

    /// <summary>
    /// Gets or sets the highlight brush.
    /// </summary>
    public IBrush? Highlight
    {
        get { return GetValue(HighlightProperty); }
        set { SetValue(HighlightProperty, value); }
    }

    /// <summary>
    /// Important.
    /// </summary>
    protected override Type StyleKeyOverride { get; } = typeof(GridSplitter);

    private void PointerMovedHandler(object? sender, PointerEventArgs e)
    {
        _originalBackground ??= Background;
        Background = Highlight;
        _timer.Start();
    }

    private void TimerHandler(object? sender, EventArgs e)
    {
        if (!IsPointerOver)
        {
            Background = _originalBackground;
            _originalBackground = null;
            _timer.Stop();
        }
    }

}
```

**Comment by timunie** [comment]
You have HighlighBrush where I guess you want to use this inside a custom style? If so, StyleKeyOverride maybe wrong here. Can you prepare a full minimal sample App to investigate this?

**Comment by kuiperzone** [comment]
Yes, will do. Give me a day or so.

In the meantime, want to know something strange? I just test built my app with NativeAOT and it seems to work in that mode.

**Comment by timunie** [comment]
> I just test built my app with NativeAOT and it seems to work in that mode.

huh? that's strange. But that makes me think it's not an Avalonia issue. So will convert it into a question for now. Awaiting your sample ;-)

**Comment by kuiperzone** [comment]
Minimal sample attached.

[Sample.zip](https://github.com/AvaloniaUI/Avalonia/files/12704200/Sample.zip)

**Reply by timunie** [reply]
I managed to make it work again. Main issues were the DispatcherTimer running after init, which is wrong in your case. So I stopped it. Also I think it's better to override OnPointerMoved. 

```cs
/// <summary>
/// Custom GridSplitter.
/// </summary>
public class CustomSplitter : GridSplitter
{
    private DispatcherTimer _timer;
    private IBrush? _originalBackground;
    
    /// <summary>
    /// Custom property.
    /// </summary>
    public static readonly StyledProperty<IBrush?> HighlightProperty =
        AvaloniaProperty.Register<CustomSplitter, IBrush?>(nameof(Highlight), Brushes.Gray);

    public CustomSplitter()
    {
        _timer = new(TimeSpan.FromMilliseconds(250), DispatcherPriority.Normal, TimerHandler);
        _timer.Stop(); // timer should not run after init (it does by default). So make sure it has stopped before leaving constructor. 
    }

    /// <summary>
    /// Gets or sets the highlight brush.
    /// </summary>
    public IBrush? Highlight
    {
        get { return GetValue(HighlightProperty); }
        set { SetValue(HighlightProperty, value); }
    }

    /// <summary>
    /// Important.
    /// </summary>
    protected override Type StyleKeyOverride => typeof(GridSplitter);

    // use override instead of adding handler. 
    protected override void OnPointerMoved(PointerEventArgs e)
    {
        _originalBackground ??= Background;
        Background = Highlight;
        _timer.Start();
        
        // needed to make base class work
        base.OnPointerMoved(e);
    }
    
    private void TimerHandler(object? sender, EventArgs e)
    {
        if (!IsPointerOver)
        {
            Background = _originalBackground;
            _originalBackground = null;
            _timer.Stop();
        }
    }
}
```

please double-check if it is the result you expected.

**Reply by kuiperzone** [reply]
Tim, That's great! I never expected that. Yes, it works fine -- I wasn't aware DispatchTimer started automatically. Thank you! Andy

**Reply by maxkatz6** [reply]
This DispatchTimer constructor overload with full parameters indeed starts the timer. It was made to be consistent with WPF.
If you don't specify handler as a parameter, then it needs a manual start.

**Comment by kuiperzone** [comment]
> > I just test built my app with NativeAOT and it seems to work in that mode.
> 
> huh? that's strange. But that makes me think it's not an Avalonia issue. So will convert it into a question for now. Awaiting your sample ;-)

Ping. Sample above. Looks like a bug from the user side.

**Reply by timunie** [reply]
will take a look later. I need some time to investigate this properly.
